tests/test_flake8_tidy_imports.py

```diff
@@ -387,6 +389,213 @@ def test_I251_from_unittest_import_mock_as(flake8_path):
+            + "foo_prime.bar_rename instead."
+        ),
+        (
+            "./example.py:3:1: I251 Banned import 'foo.bar' used - use "
+            + "foo_prime.bar_rename instead."
+        ),
+    ]
+
+
+def test_I251_from_unittest_import_wildcard_no_match(flake8_path):
+    (flake8_path / "example.py").write_text(
+        dedent(
+            """\
+            from foo.bar import Bar
+
+            Bar
+            """
+        )
+    )
+    result = flake8_path.run_flake8(
+        extra_args=["--banned-modules", "foo.baz.* = foo.baz is banned"]
+    )
+    assert result.out_lines == []
+
+
+def test_I251_from_unittest_import_wildcard_multiple_matches(flake8_path):
+    (flake8_path / "example.py").write_text(
+        dedent(
+            """\
+            import foo.baz
+            import foo.bar.bat
+
+            [foo.baz, foo.bar.bat]
+            """
+        )
+    )
+    (flake8_path / "setup.cfg").write_text(
+        default_setup_cfg
+        + dedent(
+            """\
+            banned-modules = foo.* = general
+                             foo.bar.* = specific
+            """
+        )
+    )
+    result = flake8_path.run_flake8()
+    # foo.* matches both, but we prefer the longer matching pattern
+    assert result.out_lines == [
+        "./example.py:1:1: I251 Banned import 'foo.baz' used - general.",
+        "./example.py:2:1: I251 Banned import 'foo.bar.bat' used - specific.",
+    ]
+
+
+@pytest.mark.parametrize(
+    "module, expected_message",
+    (
+        ("foo", "concrete"),
+        ("foo.bar", "structured specific"),
+        ("foo.bar.bat", "structured specific"),
+        ("foo.bat", "structured general"),
+        ("foo.baz", "unstructured first"),
+        ("foo.bar.bat.baz", "unstructured first"),
+        ("foo.quux.baz", "unstructured first"),
+    ),
+)
+def test_I251_wildcard_precedence(flake8_path, module, expected_message):
+    """
+    Check that if there are multiple possible rules matching, they are matched
+    in the expected order.
+    """
+    (flake8_path / "example.py").write_text(
+        dedent(
+            f"""\
+            import {module}
+            [{module}]
+            """
+        )
+    )
+    (flake8_path / "setup.cfg").write_text(
+        default_setup_cfg
+        + dedent(
+            """\
+            banned-modules = foo = concrete
+                             foo.* = structured general
+                             foo.bar.* = structured specific
+                             foo.*.baz = unstructured first
+                             foo.*.quux.baz = unstructured second
+            """
+        )
+    )
+    result = flake8_path.run_flake8()
+    assert result.out_lines == [
+        f"./example.py:1:1: I251 Banned import '{module}' used - {expected_message}.",
+    ]
+
+
+@pytest.mark.parametrize(
+    "banned_module, imported_module, expected",
+    (
+        # Concrete matches
```

Comment:
This tests just checks whether a rule matches a module, without the rest of the flake8 runner.